DOCS-9998-sftp-connection-editorial-kt (#1257)
mulesoft/docs-connectors: sftp/1.3/modules/ROOT/edits/sftp-connection-dynamically.adoc

= Connect to SFTP Servers Dynamically Example - Mule 4

> Many integrations require connecting to different servers based on an specific condition, such as:

Reviewer: ```suggestion
Many integrations require connecting to different servers based on a specific condition, such as:
```

> * Connecting to different invoice storage servers depending on the branch that emits an invoice.

Reviewer: ```suggestion
* Connecting to different invoice storage servers depending on the branch that emits an invoice
```

> * Connecting to different servers depending on an integration subject, such as in a multi-tenant use case.

Reviewer: ```suggestion
* Connecting to different servers depending on an integration subject, such as in a multitenant use case
```

> To accommodate these use cases, the Anypoint Connector for SFTP (SFTP Connector) global configuration element supports parameters expressions that evaluate these conditions and connect to the correct server.

Reviewer: ```suggestion
To accommodate these use cases, the Anypoint Connector for SFTP (SFTP Connector) global configuration element supports parameter expressions that evaluate these conditions and connect to the correct server.
```

The following example shows how to configure SFTP Connector to connect dynamically to multiple servers:

. An HTTP Listener source initiates the flow by reading random content posted via HTTP.

> . The File Connector Read operation loads a recipients CSV file that contains a random set of SFTP destinations, with columns such as host, user, and port.

Reviewer: column names in mono?

> . A DataWeave transformation maps the content and splits the CSV file.

Reviewer: Will your audience know what you mean by "splits the CSV file"? into what?

> . A For Each component and an SFTP Write operation write the contents into each of the SFTP destinations. On each For Each iteration, the expressions set in the SFTP global configuration element resolves to a different value and establish different connections to each of the servers.

Reviewer: per list style, break after first sentence
```suggestion
. A For Each component and an SFTP Write operation write the contents into each of the SFTP destinations. 
+
On each For Each iteration, the expressions set in the SFTP global configuration element resolve to a different value and establish different connections to each of the servers.
```

> To accomplish this example, create the Mule application, configure HTTP, File and SFTP global configuration elements, and run and test the application with curl commands:

Reviewer: "accomplish" an example? I think "run" or "test" or similar.
```suggestion
To test this example, create the Mule application; configure HTTP, File, and SFTP global configuration elements; and run and test the application with curl commands:
```

.Connect to SFTP Servers Dynamically flow
image::sftp-connection-dynamically.png[Connect to SFTP Servers Dynamically]

> == Create the Mule Application

Reviewer: Doublecheck the topic hierarchy:

= Connect... (lists actions "create the Mule application, configure HTTP, File and SFTP global configuration elements, and run and test the application with curl commands")
== Create Application
=== Configure Components
"Continue creating your Mule application..."
=== Run and Test
"After creating..."
== XML for Connecting

The Level-1 intro makes it sound like create, configure, and run and test are three peer actions.
The formatting puts configure and run and text under create.
Run and test are definitely separate (and equal) to create and configure.
XML is not mentioned in the topic set intro.

Reviewer: OK...back to copyediting! ;-)

To create the Mule flow:

. In the *Mule Palette* view, select *HTTP > Listener*.
. Drag *HTTP Listener* to the Studio canvas.
. On the *HTTP Listener* configuration screen, optionally change the value of the *Display Name* field.
. Set the *Path* field to `/multitenant`
. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of HTTP Listener in the app.
. On the *General* tab, set the following parameters:

> +

Reviewer: Do we need linespace before and after bullet lists?

* *Host*: `All Interfaces [0.0.0.0] (default)`
* *Port*: `8081`
+
. Click *OK*.
. Drag a *Set Variable* component to the right of the *HTTP Listener* source.

> . Set the *Name* field to `content`, and the *Value* field to `#[payload]`.

Reviewer: no comma
```suggestion
. Set the *Name* field to `content` and the *Value* field to `#[payload]`.
```

. Drag the File Connector *Read* operation to the right of the *Set Variable* component.

> . Set the *File Path* field to `recipients.csv`. +

Reviewer: You do need a linebreak after this...:
```suggestion
. Set the *File Path* field to `recipients.csv`. 
+
```

> This reads a CSV file that contains a random set of SFTP destinations with columns such as host, user, and port.

Reviewer: column names mono

. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of File Connector in the app.
. Accept the default settings and click *OK*.

> . On the *MIME Type* tab of the *Read* operation, for the *MIME Type* field select `application/csv`.

Reviewer: ```suggestion
. On the *MIME Type* tab of the *Read* operation, for the *MIME Type* field, select `application/csv`.
```

=== Configure Additional Components and Operations

> Continue creating your Mule application by adding additional components and operations:

Reviewer: See comments about org.

. In Studio, drag a *Transform Message* component to the right of the *Read* operation.
. In the *Output* section of the component, add the following DataWeave code to map the columns of the CSV file:
+
.DataWeave script:
[source,dataweave,linenums]
----
%dw 2.0
output application/java
---
payload map using (row = $) {
   host: row.Host,
   user: row.User,
   password: row.Password}
----
+
. Drag a *For Each* component to the right of the *Transform Message* component.

> . Drag an *SFTP Write* operation into the *For Each* component. +

Reviewer: add linebreak (If you look at this in View mode, you can see that + placed at the end of a line does not create a linebreak. You have to have the + on its own line to do that.

On each For Each iteration, the *Write* operation writes the contents into each of the SFTP destinations.
. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of SFTP Write operation in the app, and set the following parameters:
+
* *Host*: `#[payload.host]`
* *Username*: `#[payload.username]`
* *Password*: `#[payload.password]`
+
[start=6]
. Click *OK*.
. On the configuration screen, set the *Path* field to `demo.txt`.
. Set the *Content* field to `payload`.
. Drag a *Set Payload* component to the right of the *For Each* component.
. Set the *Value* field to `Multicast Ok`.

=== Run and Test your Mule Application

After creating your Mule application, run and test it:

. In Studio, save your Mule app.

> . Click the project name in *Package Explorer* and then click *Run* > *Run As* > *Mule Application*. +.

Reviewer: ```suggestion
. Click the project name in *Package Explorer* and then click *Run* > *Run As* > *Mule Application*. 
```

. Open a browser and type `+http://0.0.0.0:8081/multitenant+`. +

> Note on Mule 4 behavior for Mule 3 users:

Reviewer: ```suggestion
Note about Mule 4 behavior for Mule 3 users:
```
OR
```suggestion
If you use Mule 3, note that Mule 4 behaves in the following ways:
```

> * The information posted through the HTTP Listener source is written to each SFTP site multiple times. Because the listener makes use of the xref:mule-runtime::streaming-about.adoc[repeatable streams feature], you do not need to worry about consuming the stream multiple times.

Reviewer: linebreak
```suggestion
* The information posted through the HTTP Listener source is written to each SFTP site multiple times. 
+
Because the listener makes use of the xref:mule-runtime::streaming-about.adoc[repeatable streams feature], you do not need to worry about consuming the stream multiple times.
```

> * The For Each component automatically goes through each line of the CSV file. In Mule 3, you first needed to transform the CSV file into a Java structure, but because Mule 4 is now Java agnostic, this works out-of-the-box.

Reviewer: linebreak
```suggestion
* The For Each component automatically goes through each line of the CSV file. 
+
In Mule 3, you first needed to transform the CSV file into a Java structure, but because Mule 4 is now Java agnostic, this works out of the box.
```

> When reading or listing files, you might be interested in the file's metadata (for example, the file name, full path, size, timestamp, and so on). The connector uses the Mule Message attributes to access this information.

Reviewer: Per style ref., GLOBALLY avoid "and so on":
```suggestion
When reading or listing files, you might be interested in the file's metadata (for example, the file name, full path, size, and timestamp). The connector uses the Mule Message attributes to access this information.
```
Why is "message" init. capped?
